### PR TITLE
Y0 restart

### PR DESCRIPTION
**Pull request by MTCam**
This adds an experimental restart capability to the Y0 driver.  Note the use of the `my_checkpoint` function in the driver that allows the user to add functionality to the checkpointing. 

edit: Sorry for the deep diff.  This PR also merged int current master and @inducer's restart pilot.  The relevant bits are [where it reads](https://github.com/illinois-ceesd/mirgecom/pull/178/files#r543886682), and [where it writes](https://github.com/illinois-ceesd/mirgecom/pull/178/files#r543885774).

### CODE REVIEW

File: examples/y0euler_restart.py [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+                    "state": obj_array_vectorize(actx.to_numpy, flatten(state)),
+                    "t": t,
+                    "step": step,
+                    "global_nelements": global_nelements,
+                    "num_parts": nparts,
+                    }, f)
+        return sim_checkpoint(discr=discr, visualizer=visualizer, eos=eos,
+                              q=state, vizname=casename,
+                              step=step, t=t, dt=dt, nstatus=nstatus,
+                              nviz=nviz, constant_cfl=constant_cfl, comm=comm)
```
Comment (MTCam) [inline examples/y0euler_restart.py:303]: Restarts are written here when the time stepping loop calls `checkpoint`

File: examples/y0euler_restart.py [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+        if rank == 0:
+            logging.info("Initializing soln.")
+        current_state = bulk_init(0, nodes, eos=eos)
+    else:
+        current_t = restart_data["t"]
+        current_step = restart_step
+
+        current_state = unflatten(
+            actx, discr.discr_from_dd("vol"),
+            obj_array_vectorize(actx.from_numpy, restart_data["state"]))
```
Comment (MTCam) [inline examples/y0euler_restart.py:260]: Restart data is read here on restart.

**Comment by inducer** [comment]
- Could you just push the merges to the target branch so that the diff is exactly what needs to be reviewed?
- CI fails

**Comment by MTCam** [comment]
> * Could you just push the merges to the target branch so that the diff is exactly what needs to be reviewed?
> * CI fails

I have done the merge thing in the base branch.   Converting this to a WIP until @anderson2981 has it working on Y0.

**Pull request merged**